Internal compiler error: Encountered ambiguity during codegen

I'm not exactly sure what's happening here, but It seems as though calling a trait as a function, with `Box<Any>` leads to a codegen bug.

__I tried this code ([playground link](https://play.integer32.com/?version=stable&mode=debug&edition=2018&gist=a7184ea2614bf9209c486de2e90ca5cb))__:
```
pub trait Foo: Sized {
    const SIZE: usize;
    
    fn new(slice: &[u8; Foo::SIZE]) -> Self;
}

pub struct Bar<T: ?Sized>(T);

impl Bar<[u8]> {
    const SIZE: usize = 32;

    fn new(slice: &[u8; Self::SIZE]) -> Self {
        Foo(Box::new(*slice))
    }
}

fn main() {
    println!("Hello, world!");
}
```
__I expected to see this happen__:
That the compiler throws `error[E0423]: expected function, found trait Foo`, which it does, but without an internal compiler error.

__Instead, this happened__:
```
error: internal compiler error: src/librustc/traits/codegen/mod.rs:63: Encountered ambiguity selecting `Binder(<[type error] as Foo>)` during codegen, presuming due to overflow

thread 'main' panicked at 'Box<Any>', src/librustc_errors/lib.rs:600:9
```

## Meta

`note: rustc 1.32.0 (9fda7c223 2019-01-16) running on x86_64-unknown-linux-gnu`

`note: compiler flags: -C codegen-units=1 -C debuginfo=2 --crate-type bin`

Backtrace:
```
thread 'main' panicked at 'Box<Any>', src/librustc_errors/lib.rs:600:9
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:59
             at src/libstd/panicking.rs:211
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:227
   4: rustc::util::common::panic_hook
   5: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:495
   6: std::panicking::begin_panic
   7: rustc_errors::Handler::bug
   8: rustc::util::bug::opt_span_bug_fmt::{{closure}}
   9: rustc::ty::context::tls::with_opt::{{closure}}
  10: rustc::ty::context::tls::with_context_opt
  11: rustc::ty::context::tls::with_opt
  12: rustc::util::bug::opt_span_bug_fmt
  13: rustc::util::bug::bug_fmt
  14: rustc::ty::context::tls::with_related_context
  15: rustc::infer::InferCtxtBuilder::enter
  16: rustc::traits::codegen::codegen_fulfill_obligation
  17: rustc::ty::query::__query_compute::codegen_fulfill_obligation
  18: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors<'tcx> for rustc::ty::query::queries::codegen_fulfill_obligation<'tcx>>::compute
  19: rustc::dep_graph::graph::DepGraph::with_task_impl
  20: <rustc::ty::query::plumbing::JobOwner<'a, 'tcx, Q>>::start
  21: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::force_query_with_job
  22: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
  23: rustc::ty::instance::Instance::resolve
  24: <rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, M>>::resolve
  25: rustc_mir::interpret::operand::<impl rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, M>>::const_value_to_op
  26: rustc_mir::interpret::operand::<impl rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, M>>::eval_operand
  27: rustc_mir::interpret::step::<impl rustc_mir::interpret::eval_context::EvalContext<'a, 'mir, 'tcx, M>>::run
  28: rustc_mir::const_eval::eval_body_using_ecx
  29: rustc_mir::const_eval::const_eval_raw_provider
  30: rustc::ty::query::__query_compute::const_eval_raw
  31: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors<'tcx> for rustc::ty::query::queries::const_eval_raw<'tcx>>::compute
  32: rustc::dep_graph::graph::DepGraph::with_task_impl
  33: <rustc::ty::query::plumbing::JobOwner<'a, 'tcx, Q>>::start
  34: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::force_query_with_job
  35: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
  36: rustc::ty::query::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::const_eval_raw
  37: rustc_mir::const_eval::const_eval_provider
  38: rustc::ty::query::__query_compute::const_eval
  39: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors<'tcx> for rustc::ty::query::queries::const_eval<'tcx>>::compute
  40: rustc::dep_graph::graph::DepGraph::with_task_impl
  41: <rustc::ty::query::plumbing::JobOwner<'a, 'tcx, Q>>::start
  42: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::force_query_with_job
  43: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
  44: <rustc::traits::project::AssociatedTypeNormalizer<'a, 'b, 'gcx, 'tcx> as rustc::ty::fold::TypeFolder<'gcx, 'tcx>>::fold_const
  45: rustc::ty::structural_impls::<impl rustc::ty::fold::TypeFoldable<'tcx> for &'tcx rustc::ty::TyS<'tcx>>::super_fold_with
  46: <rustc::traits::project::AssociatedTypeNormalizer<'a, 'b, 'gcx, 'tcx> as rustc::ty::fold::TypeFolder<'gcx, 'tcx>>::fold_ty
  47: rustc::ty::structural_impls::<impl rustc::ty::fold::TypeFoldable<'tcx> for &'tcx rustc::ty::TyS<'tcx>>::super_fold_with
  48: <rustc::traits::project::AssociatedTypeNormalizer<'a, 'b, 'gcx, 'tcx> as rustc::ty::fold::TypeFolder<'gcx, 'tcx>>::fold_ty
  49: <smallvec::SmallVec<A> as core::iter::traits::FromIterator<<A as smallvec::Array>::Item>>::from_iter
  50: rustc::ty::fold::TypeFoldable::fold_with
  51: rustc::traits::project::normalize
  52: rustc::infer::InferCtxt::partially_normalize_associated_types_in
  53: rustc::ty::context::tls::with_related_context
  54: rustc::infer::InferCtxtBuilder::enter
  55: rustc_typeck::check::wfcheck::check_associated_item
  56: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors<'tcx> for rustc::ty::query::queries::check_trait_item_well_formed<'tcx>>::compute
  57: rustc::dep_graph::graph::DepGraph::with_task_impl
  58: <rustc::ty::query::plumbing::JobOwner<'a, 'tcx, Q>>::start
  59: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::force_query_with_job
  60: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query
  61: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::ensure_query
  62: rustc::session::Session::track_errors
  63: rustc::util::common::time
  64: rustc_typeck::check_crate
  65: rustc::ty::context::tls::enter_context
  66: <std::thread::local::LocalKey<T>>::with
  67: rustc::ty::context::TyCtxt::create_and_enter
  68: rustc_driver::driver::compile_input
  69: rustc_driver::run_compiler_with_pool
  70: <scoped_tls::ScopedKey<T>>::set
  71: rustc_driver::run_compiler
  72: rustc_driver::monitor::{{closure}}
  73: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:102
  74: rustc_driver::run
  75: rustc_driver::main
  76: std::rt::lang_start::{{closure}}
  77: std::panicking::try::do_call
             at src/libstd/rt.rs:59
             at src/libstd/panicking.rs:310
  78: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:102
  79: std::rt::lang_start_internal
             at src/libstd/panicking.rs:289
             at src/libstd/panic.rs:398
             at src/libstd/rt.rs:58
  80: main
  81: __libc_start_main
  82: <unknown>
query stack during panic:
#0 [codegen_fulfill_obligation] checking if `Foo` fulfills its obligations
#1 [const_eval_raw] const-evaluating `Foo::new::{{constant}}`
#2 [const_eval] const-evaluating + checking `Foo::new::{{constant}}`
#3 [check_trait_item_well_formed] processing `Foo::new`
end of query stack
```

--- a/src/test/ui/issues/issue-50571.rs
+++ b/src/test/ui/issues/issue-50571.rs
@@ -1,0 +1,6 @@
+trait Foo {
+    fn foo([a, b]: [i32; 2]) {}
+    //~^ ERROR: patterns aren't allowed in methods without bodies
+}
+
+fn main() {}

--- a/src/test/ui/issues/issue-50571.stderr
+++ b/src/test/ui/issues/issue-50571.stderr
@@ -1,0 +1,13 @@
+error[E0642]: patterns aren't allowed in methods without bodies
+  --> $DIR/issue-50571.rs:2:12
+   |
+LL |     fn foo([a, b]: [i32; 2]) {}
+   |            ^^^^^^
+help: give this argument a name or use an underscore to ignore it
+   |
+LL |     fn foo(_: [i32; 2]) {}
+   |            ^
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0642`.

--- a/src/test/ui/issues/issue-58022.rs
+++ b/src/test/ui/issues/issue-58022.rs
@@ -1,0 +1,18 @@
+pub trait Foo: Sized {
+    const SIZE: usize;
+
+    fn new(slice: &[u8; Foo::SIZE]) -> Self;
+    //~^ ERROR: type annotations needed: cannot resolve `_: Foo`
+}
+
+pub struct Bar<T: ?Sized>(T);
+
+impl Bar<[u8]> {
+    const SIZE: usize = 32;
+
+    fn new(slice: &[u8; Self::SIZE]) -> Self {
+        Foo(Box::new(*slice)) //~ ERROR: expected function, found trait `Foo`
+    }
+}
+
+fn main() {}

--- a/src/test/ui/issues/issue-58022.stderr
+++ b/src/test/ui/issues/issue-58022.stderr
@@ -1,0 +1,19 @@
+error[E0423]: expected function, found trait `Foo`
+  --> $DIR/issue-58022.rs:14:9
+   |
+LL |         Foo(Box::new(*slice))
+   |         ^^^ not a function
+
+error[E0283]: type annotations needed: cannot resolve `_: Foo`
+  --> $DIR/issue-58022.rs:4:25
+   |
+LL |     const SIZE: usize;
+   |     ------------------ required by `Foo::SIZE`
+LL | 
+LL |     fn new(slice: &[u8; Foo::SIZE]) -> Self;
+   |                         ^^^^^^^^^
+
+error: aborting due to 2 previous errors
+
+Some errors have detailed explanations: E0283, E0423.
+For more information about an error, try `rustc --explain E0283`.

--- a/src/test/ui/issues/issue-58344.rs
+++ b/src/test/ui/issues/issue-58344.rs
@@ -1,0 +1,50 @@
+use std::ops::Add;
+
+trait Trait<T> {
+    fn get(self) -> T;
+}
+
+struct Holder<T>(T);
+
+impl<T> Trait<T> for Holder<T> {
+    fn get(self) -> T {
+        self.0
+    }
+}
+
+enum Either<L, R> {
+    Left(L),
+    Right(R),
+}
+
+impl<L, R> Either<L, R> {
+    fn converge<T>(self) -> T where L: Trait<T>, R: Trait<T> {
+        match self {
+            Either::Left(val) => val.get(),
+            Either::Right(val) => val.get(),
+        }
+    }
+}
+
+fn add_generic<A: Add<B>, B>(lhs: A, rhs: B) -> Either<
+    impl Trait<<A as Add<B>>::Output>,
+    impl Trait<<A as Add<B>>::Output>
+> {
+    if true {
+        Either::Left(Holder(lhs + rhs))
+    } else {
+        Either::Right(Holder(lhs + rhs))
+    }
+}
+
+fn add_one(
+    value: u32,
+) -> Either<impl Trait<<u32 as Add<u32>>::Output>, impl Trait<<u32 as Add<u32>>::Output>> {
+    //~^ ERROR: the trait bound `impl Trait<<u32 as std::ops::Add>::Output>: Trait<u32>`
+    //~| ERROR: the trait bound `impl Trait<<u32 as std::ops::Add>::Output>: Trait<u32>`
+    add_generic(value, 1u32)
+}
+
+pub fn main() {
+    add_one(3).converge();
+}

--- a/src/test/ui/issues/issue-58344.stderr
+++ b/src/test/ui/issues/issue-58344.stderr
@@ -1,0 +1,19 @@
+error[E0277]: the trait bound `impl Trait<<u32 as std::ops::Add>::Output>: Trait<u32>` is not satisfied
+  --> $DIR/issue-58344.rs:42:13
+   |
+LL | ) -> Either<impl Trait<<u32 as Add<u32>>::Output>, impl Trait<<u32 as Add<u32>>::Output>> {
+   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Trait<u32>` is not implemented for `impl Trait<<u32 as std::ops::Add>::Output>`
+   |
+   = note: the return type of a function must have a statically known size
+
+error[E0277]: the trait bound `impl Trait<<u32 as std::ops::Add>::Output>: Trait<u32>` is not satisfied
+  --> $DIR/issue-58344.rs:42:52
+   |
+LL | ) -> Either<impl Trait<<u32 as Add<u32>>::Output>, impl Trait<<u32 as Add<u32>>::Output>> {
+   |                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Trait<u32>` is not implemented for `impl Trait<<u32 as std::ops::Add>::Output>`
+   |
+   = note: the return type of a function must have a statically known size
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0277`.

